sql: Add the `CREATE SCHEDULE FOR BACKUP` grammar

miretskiy opened the pull request:
Informs #28266

Add a new CREATE SCHEDULE FOR BACKUP statement grammar which allows
the specification of the BACKUP that needs to be executed periodically.

Perform full cluster backup daily:

```
CREATE SCHEDULE 'full cluster'
FOR BACKUP OF CLUSTER TO 'scheme://uri/for/backup'
RECURRING '@daily'
```

The frequency and the time of the schedule can be specified using
crontab format.

The RECURRING expression specifies when we backup.  By default these are
incremental backups that capture changes since the last backup,
writing to the "current" backup.  We will choose a resonable default for
the frequency of full backups.  However, the users may also specify
when and how often to run full backups explicitly.

For example, the following will perform incremental backups daily,
and full backup on Sundays, at 4 am (EST)

```
CREATE SCHEDULE 'full cluster'
FOR BACKUP OF CLUSTER TO 'scheme://uri/for/backup'
RECURRING '@daily'
FULL BACKUP 'TZ=America/New_York Sun * * 4 0'
```

The full list of options supported by this statement:
```
  CREATE SCHEDULE [<description>] FOR BACKUP [OF <targets>] TO  <location...>
  [WITH <backup_option> = [<value>] [, ...]]
  RECURRING [crontab|NEVER] [FULL BACKUP <crontab|ALWAYS>]
  [WITH EXPERIMENTAL SCHEDULE OPTIONS <schedule_option>[= <value>] [, ...] ]
```

Release Note (sql change): Add CREATE SCHEDULE FOR BACKUP grammar.
This statement can be used to specify the recurring backup schedules.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50631)
<!-- Reviewable:end -->

rohany reviewed (changes requested):
I feel like there is a good amount of extra logic/validation happening in the parser that should be done in later validation passes over the statement.

rohany commented on pkg/ccl/jobsccl/create_schedule_stmt.go:
```diff
@@ -0,0 +1,65 @@
+	resultsCh <- tree.Datums{
+		tree.NewDInt(tree.DInt(0)),
+		tree.DNull,
+		tree.NewDString(fmtCtx.String()),
```
[nit]: You can just do `tree.AsString(schedule)` here.

rohany commented on pkg/sql/parser/sql.y:
```diff
@@ -2028,6 +2048,192 @@ backup_stmt:
+| backup_schedule_options APPENDING CHANGES EVERY interval_value
+  {
+    opts := $1.backupScheduleOptions()
+    if opts.ChangeCapturePeriod != nil {
```
If you made this rule `schedule_options APPENDING CHANGES EVERY interval_value`, then it wouldn't be possible to have multiple `APPENDING CHANGES` options. But why is this a separate case than from the case below?

miretskiy reviewed (commented):
There we many changes done to this PR, mostly due to the discussion in https://docs.google.com/document/d/1TeYYEGKomTRiWIZf_R0tYqnOEsnC45HM_2p5CxHXF_0/edit?ts=5ef4e10e#

Parsing tests added.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-:-MBB-q6BAYfRP-I_Y3ea:b-6enyrd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rohany)

___
*[pkg/ccl/jobsccl/create_schedule_stmt.go, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MAcU-uDjxg34Qw-mcF0-r1-39:-MAg183I9eG30LKheULI:b-97gzxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/af3cea14120710936a81a27e6583557307afc17a/pkg/ccl/jobsccl/create_schedule_stmt.go#L39)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit]: You can just do `tree.AsString(schedule)` here.
</blockquote></details>

Thanks.  Done.
___
*[pkg/sql/parser/sql.y, line 2107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MAcU-uDjxg34Qw-mcEz-r1-2107:-MBB-m5oB8dYf5oc2b3m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af3cea14120710936a81a27e6583557307afc17a/pkg/sql/parser/sql.y#L2107)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

All these changes need corresponding tests in `parse_test.go`
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 2146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MAcU-uDjxg34Qw-mcEz-r1-2146:-MAg7JehEnS1pSNvzpEz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af3cea14120710936a81a27e6583557307afc17a/pkg/sql/parser/sql.y#L2146)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

If you made this rule `schedule_options APPENDING CHANGES EVERY interval_value`, then it wouldn't be possible to have multiple `APPENDING CHANGES` options. But why is this a separate case than from the case below?
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 2163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MAgbbq8BH1GD0PPaK9y:-MAgbbq8BH1GD0PPaK9z:b-m6bdkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/af3cea14120710936a81a27e6583557307afc17a/pkg/sql/parser/sql.y#L2163)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
>     opts := $1.scheduleOptions()
>     if opts.FirstRun != nil {
>       return setErr(sqllex,  errors.New("STARTING option specified multiple times"))
>     }
>     opts.FirstRun = $3.expr()
> ```
</details>

per our discussion: moved this validation to schedule.go


<!-- Sent from Reviewable.io -->

rohany reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50631)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, @pbardea, and @rohany)

___
*[pkg/ccl/jobsccl/create_schedule_stmt_test.go, line 23 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MBEuEouDSnupxje0Ge6:-MBEuEov3_NnDUtB1k-D:bhtfy11) ([raw file](https://github.com/cockroachdb/cockroach/blob/59e5dcfadae7a70d87b1973db56a4e2bd70da317/pkg/ccl/jobsccl/create_schedule_stmt_test.go#L23)):*
> ```Go
> )
> 
> func TestCreateNoopSchedule(t *testing.T) {
> ```

If this is just a parsing test, then we have those already -- do we need this test?
___
*[pkg/sql/parser/sql.y, line 2245 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MBEuqUB5k1bc_9sDa74:-MBEuqUB5k1bc_9sDa75:bhma1zc) ([raw file](https://github.com/cockroachdb/cockroach/blob/59e5dcfadae7a70d87b1973db56a4e2bd70da317/pkg/sql/parser/sql.y#L2245)):*
> ```Yacc
>   }
> 
> sconst_or_placeholder:
> ```

Why do you need this rule? Wouldn't just taking an expr where you are using it be fine? I'm guessing you don't want to validate it later


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50631)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @miretskiy, @pbardea, and @rohany)

___
*[pkg/sql/parser/sql.y, line 2245 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MBEuqUB5k1bc_9sDa74:-MBF61ipCviTKIbZ61ux:b92lwpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/59e5dcfadae7a70d87b1973db56a4e2bd70da317/pkg/sql/parser/sql.y#L2245)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Why do you need this rule? Wouldn't just taking an expr where you are using it be fine? I'm guessing you don't want to validate it later
</blockquote></details>

I wanted to use string_or_placeholder (in cron_expr) rule.  Alas, string_or_placeholder also matches unreserved keywords -- which results in reduce/reduce conflict.


<!-- Sent from Reviewable.io -->

rohany commented:
> If this is just a parsing test, then we have those already -- do we need this test?

LGTM, but bumping this question:

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50631)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rohany)

___
*[pkg/ccl/jobsccl/create_schedule_stmt_test.go, line 23 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MBEuEouDSnupxje0Ge6:-MBWUGQYCbkB_ohsH5l5:b-w29r43) ([raw file](https://github.com/cockroachdb/cockroach/blob/59e5dcfadae7a70d87b1973db56a4e2bd70da317/pkg/ccl/jobsccl/create_schedule_stmt_test.go#L23)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

If this is just a parsing test, then we have those already -- do we need this test?
</blockquote></details>

Good point.  Removing


<!-- Sent from Reviewable.io -->

Commit:
sql: Add the `CREATE SCHEDULE FOR BACKUP` statement

Add a new `CREATE SCHEDULE FOR BACKUP` statement which allows
the specification of the BACKUP that needs to be executed periodically.

Perform full cluster backup daily:
```
CREATE SCHEDULE 'full cluster'
FOR BACKUP OF CLUSTER TO 'scheme://uri/for/backup'
RECURRING '@daily'
```

The frequency and the time of the schedule can be specified using
crontab format.

It is possible to request both full as well as delta backups.
The following will perform full table backup daily (midnight), but
produce delta backups every 15 minutes:

```
CREATE SCHEDULE 'table user_data backup'
FOR BACKUP OF TABLE user_data TO 'scheme://uri/for/backup'
RECURRING '@daily'
APPENDING CHANGES EVERY INTERVAL '15 minutes'
```

The full list of options supported by this statement:
```
  CREATE SCHEDULE [<description>] FOR BACKUP [OF <targets>] TO  <location...>
    [WITH <backup_option> = [<value>] [, ...]]
    [STARTING <time_expr>] [RECURRING <crontab>]
    [ON EXECUTION FAILURE <PAUSE|RETRY|CONTINUE>]
    [APPENDING CHANGES EVERY INTERVAL <interval>]
    [ON PREVIOUS RUNNING (START|SKIP|WAIT)]
```

Release Note (sql change): Add `CREATE SCHEDULE FOR BACKUP` statement.
This statement can be used to specify the recurring backup schedules.

pbardea commented on pkg/sql/parser/parse_test.go:
```diff
@@ -1431,6 +1431,12 @@ func TestParse(t *testing.T) {
+		{`CREATE SCHEDULE FOR BACKUP TABLE foo, bar, buz TO 'bar' RECURRING '@daily' FULL BACKUP ALWAYS`},
+		{`CREATE SCHEDULE FOR BACKUP TABLE foo, bar, buz TO 'bar' RECURRING '@daily' FULL BACKUP '@weekly'`},
+		{`CREATE SCHEDULE FOR BACKUP TABLE foo, bar, buz TO 'bar' WITH revision_history RECURRING '@daily' FULL BACKUP '@weekly'`},
+		{`CREATE SCHEDULE FOR BACKUP TO 'bar' WITH revision_history RECURRING '@daily' FULL BACKUP '@weekly' WITH EXPERIMENTAL SCHEDULE OPTIONS foo = 'bar'`},
```
Might as well also add a schedule description in here too?

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50631)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @rohany)

___
*[pkg/sql/parser/parse_test.go, line 1439 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/50631#-MBBVy5esWKZ6QTC3fIe-r3-1439:-MB_rrPNDJXQSOaLS3pO:b1r0t6z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f947aa695e8bbf2dbd220f54b06e19e123fd5765/pkg/sql/parser/parse_test.go#L1439)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Might as well also add a schedule description in here too?
</blockquote></details>

Ooops -- Just noticed; adding


<!-- Sent from Reviewable.io -->

miretskiy commented:
bors r+
